fotoapparat/src/main/java/io/fotoapparat/Fotoapparat.kt

```diff
@@ -0,0 +1,247 @@
+import io.fotoapparat.configuration.Configuration
+import io.fotoapparat.error.onMainThread
+import io.fotoapparat.exception.camera.CameraException
+import io.fotoapparat.hardware.Device
+import io.fotoapparat.hardware.display.Display
+import io.fotoapparat.hardware.execute
+import io.fotoapparat.hardware.executeTask
+import io.fotoapparat.hardware.orientation.OrientationSensor
+import io.fotoapparat.log.Logger
+import io.fotoapparat.log.none
+import io.fotoapparat.parameter.ScaleType
+import io.fotoapparat.parameter.camera.CameraParameters
+import io.fotoapparat.result.*
+import io.fotoapparat.routine.camera.bootStart
+import io.fotoapparat.routine.camera.shutDown
+import io.fotoapparat.routine.camera.switchCamera
+import io.fotoapparat.routine.camera.updateDeviceConfiguration
+import io.fotoapparat.routine.capability.getCapabilities
+import io.fotoapparat.routine.focus.focus
+import io.fotoapparat.routine.parameter.getCurrentParameters
+import io.fotoapparat.routine.photo.takePhoto
+import io.fotoapparat.routine.zoom.updateZoomLevel
+import io.fotoapparat.selector.back
+import io.fotoapparat.selector.external
+import io.fotoapparat.selector.firstAvailable
+import io.fotoapparat.selector.front
+import io.fotoapparat.view.CameraRenderer
+import java.util.concurrent.FutureTask
+
+/**
+ * Camera. Takes pictures.
+ */
+class Fotoapparat
+@JvmOverloads constructor(
+        context: Context,
+        view: CameraRenderer,
+        lensPosition: Collection<LensPosition>.() -> LensPosition? = firstAvailable(
+                back(),
+                front(),
+                external()
+        ),
+        scaleType: ScaleType = ScaleType.CenterCrop,
+        cameraConfiguration: CameraConfiguration = CameraConfiguration.default(),
+        cameraErrorCallback: ((CameraException) -> Unit) = {},
+        private val logger: Logger = none()
+) {
+
+    private val mainThreadErrorCallback = cameraErrorCallback.onMainThread()
+
+    private val display = Display(context)
+
+    private val device = Device(
+            cameraRenderer = view,
+            logger = logger,
+            display = display,
+            scaleType = scaleType,
+            initialLensPositionSelector = lensPosition,
+            initialConfiguration = cameraConfiguration
+    )
+
+    private val orientationSensor = OrientationSensor(
+            context = context,
+            device = device
+    )
+
+    init {
+        logger.recordMethod()
+    }
+
+    /**
+     * Starts camera.
+     *
+     * @throws IllegalStateException If the camera has already started.
+     */
+    fun start() {
+        logger.recordMethod()
+
+        execute {
+            device.bootStart(
+                    orientationSensor = orientationSensor,
+                    mainThreadErrorCallback = mainThreadErrorCallback
+            )
+        }
+    }
+
+    /**
+     * Stops camera.
+     *
+     * @throws IllegalStateException If the camera has not started.
+     */
+    fun stop() {
```

Comment:
The caller of stop() is not able to catch the IllegalStateException thrown if the camera has not been started. Can you update this to pass the exception back to the caller?

Reply:
This is by design. You must never call stop if you have not started the camera. Similar to an activity's lifecycle.

This lifecycle helps mental sanity. E.g. if you have a bug and call accidentally 2 times `stop()` or if e.g. you had forgotten to call `start()`, then if stop was optionally no-op and hiding exceptions you would scratch your head over those bugs. This way they are exposed.

Reply:
Ok. By not propagating the IllegalStateException to caller, you're forcing users of the library to maintain their own flag of whether start() was called which feels less safe to me. I don't quite understand the logic of throwing an uncatchable fatal exception for a function that could just be a no-op if it's already stopped.

Reply:
Yes you are right. The docs & what I explained above is the desired logic. Due to the thread there it's not propagating the exception (incorrectly) making impossible to catch 👍